src/canvasdrawer.js

```diff
@@ -0,0 +1,1151 @@
+
+    /**
+     * @returns {Boolean} true if canvas is supported by the browser, otherwise false
+     */
+    static isSupported(){
+        return $.supportsCanvas;
+    }
+
+    getType(){
+        return 'canvas';
+    }
+
+    /**
+     * create the HTML element (e.g. canvas, div) that the image will be drawn into
+     * @returns {Element} the canvas to draw into
+     */
+    createDrawingElement(){
+        let canvas = $.makeNeutralElement("canvas");
+        let viewportSize = this._calculateCanvasSize();
+        canvas.width = viewportSize.x;
+        canvas.height = viewportSize.y;
+        return canvas;
+    }
+
+    /**
+     * Draws the TiledImages
+     */
+    draw(tiledImages) {
+        var _this = this;
+        this._prepareNewFrame(); // prepare to draw a new frame
+        tiledImages.forEach(function(tiledImage){
+            if (tiledImage.opacity !== 0 || tiledImage._preload) {
+                // _this._updateViewportWithTiledImage(tiledImage);
+                _this._drawTiles(tiledImage);
+            }
+        });
+    }
+
+    /**
+     * @returns {Boolean} True - rotation is supported.
+     */
+    canRotate() {
+        return true;
+    }
+
+    /**
+     * Destroy the drawer (unload current loaded tiles)
+     */
+    destroy() {
+        //force unloading of current canvas (1x1 will be gc later, trick not necessarily needed)
+        this.canvas.width  = 1;
+        this.canvas.height = 1;
+        this.sketchCanvas = null;
+        this.sketchContext = null;
+    }
+
+    /**
+     * @returns {Boolean} Whether this drawer requires enforcing minimum tile overlap to avoid showing seams.
+     */
+    minimumOverlapRequired() {
+       return true;
+    }
+
+
+    /**
+     * Turns image smoothing on or off for this viewer. Note: Ignored in some (especially older) browsers that do not support this property.
+     *
+     * @function
+     * @param {Boolean} [imageSmoothingEnabled] - Whether or not the image is
+     * drawn smoothly on the canvas; see imageSmoothingEnabled in
+     * {@link OpenSeadragon.Options} for more explanation.
+     */
+    setImageSmoothingEnabled(imageSmoothingEnabled){
+        this._imageSmoothingEnabled = imageSmoothingEnabled;
+        this._updateImageSmoothingEnabled(this.context);
+        this.viewer.forceRedraw();
+    }
+
+    /**
+     * Draw a rectangle onto the canvas
+     * @param {OpenSeadragon.Rect} rect
+     */
+    drawDebuggingRect(rect) {
+        var context = this.context;
+        context.save();
+        context.lineWidth = 2 * $.pixelDensityRatio;
+        context.strokeStyle = this.debugGridColor[0];
+        context.fillStyle = this.debugGridColor[0];
+
+        context.strokeRect(
+            rect.x * $.pixelDensityRatio,
+            rect.y * $.pixelDensityRatio,
+            rect.width * $.pixelDensityRatio,
+            rect.height * $.pixelDensityRatio
+        );
+
+        context.restore();
+    }
+
+    /**
+     * @private
+     * @inner
```

Comment:
`@inner` doesn't seem relevant here? https://jsdoc.app/tags-inner Let me know if I misunderstand.

Reply:
I don't actually know. I was trying to maintain a similar doc formatting to that used in `TiledImage` (where a lot of this code came from). Many of those use comments us `@inner`.

Reply:
Cool. I'm guessing it was a mistake in `TiledImage`. Let's not use them.

Reply:
Removed